HostData CR for host inspection data (#208)
metal3-io/metal3-docs: design/baremetal-operator/configmap_for_host_inspection_data.md

# configMap for host inspection data

## Status

implementable

## Summary

Write host inspection data into ConfigMap.

## Motivation

Currently inspection data is written to baremetalhost status subresource. When
pivoting to a target cluster, with *`clusterctl move`*, status subresource is not
copied with the host, because from Cluster-API perspective, it's controllers'
job to rebuild the status in the target cluster. Also, according to the [Kubernetes API convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status), status
is expected to be reconstructed by controllers if it's lost for some reason.
Baremetal Operator (BMO) is capable of reconstructing host status by re-inspecting
the host, but it will force the already provisioned host to go through another
round of provisioning, which eventually reboots the host after IPA steps are
completed. To have the inspection data available in the target cluster and to

> avoid unnecessary provisioning after pivoting, we normally copy inspection data

Reviewer: ```suggestion
avoid unnecessary provisioning after pivoting, we currently copy inspection data
```

Author: thanks, fixed

reference to the owner object BareMetalHost. The data part of ConfigMap will contain
the same hardware information as currently under `status.hardware` of BareMetalHost.

### Pivoting

[`clusterctl.cluster.x-k8s.io/move-hierarchy`](https://github.com/kubernetes-sigs/cluster-api/blob/0e6428151f2a388b98d1e220a698b85940a5f290/cmd/clusterctl/api/v1alpha3/labels.go#L38) label should be set on BareMetalHost CRD before issuing
`clusterctl move` so that the entire hierarchy (i.e. including secrets, ConfigMaps)
is pivoted to the target cluster. Having inspection data in a ConfigMap with
ownerReferences ensures that the inspection data is available in the target cluster.

### Implementation Details/Notes/Constraints

The life of a ConfigMap is tightly coupled to the corresponding BareMetalHost.
Creating or deleting a BareMetalHost will result in creation or deletion of a
ConfigMap respectively.

**Note:** [cross-namespace owner references are not allowed](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications). Thus, ConfigMap and BareMetalHost
must be in the same namespace.

- BareMetalHost is created

> - Operator creates a ConfigMap with the same name and namespace as BareMetalHost

Reviewer 2: I don't think it's safe to require the same name here, because ConfigMaps can be used for anything; this could easily collide with something the user is doing.

    and sets ownerReferences

> - Operator writes name of the ConfigMap to `spec.hardwareInfo` of the BareMetalHost

Reviewer 2: This kind of contradicts the previous step. If the name is always the same there is no need to write it in the Spec.

- Provisioning requested, operator fills out the ConfigMap with the host
    inspection data
- Deprovisioning requested, operator cleans up the inspection data from the
    ConfigMap
- BareMetalHost deletion requested, operator deletes the ConfigMap and BareMetalHost

#### BareMetalHost API changes

> - `status.hardware` field should be removed ⚠️

Reviewer 2: We cannot do this without bumping the BMH CRD version.
We need to get to a point soon where we can do that, but in the meantime it would be good to have a plan for what to do in the meantime (populate both?).

- `spec.hardwareInfo` field should be added, to keep the name of ConfigMap as a
    reference

### Risks and Mitigations

If an unintended user gains an access to the cluster and deletes the ConfigMap,
BMO will notice the absence of the ConfigMap in the next reconciliation loop,
and will try to recreate it. This will potentially trigger another provisioning.

### Work Items

- Remove `hardware` field from the Status of BareMetalHost API
- Add `hardwareInfo` to the Spec of BareMetalHost API
- Modify the operator code
- Clean up statusAnnotation related code from BMO and CAPM3
- Add unit tests
- Update documentation
- Update dependencies in Cluster-api-provider-metal3, metal3-dev-env GitHub
    repositories
- Give a notice to the community in advance about the change, because some
    external code bases might have tight dependencies

### Dependencies

None

### Test Plan

Unit tests should be added for new functions and integration test scripts
should be updated.

### Upgrade / Downgrade Strategy

From a user perspective, this feature doesn't introduce any breaking changes
to existing deployments, except when there is reliance on the `status.hardware`

> field from external code base.

Reviewer 2: I don't agree here. Removing the HardwareDetails field is definitely a breaking change, and not populating it any more would be a breaking change for consumers of the API.

### Version Skew Strategy

I can not foresee any version skew issues.

## Drawbacks

> None. Is there anything you can imagine?

Reviewer 2: Loss of schema information, inability to set RBAC independently of other ConfigMaps.

## Alternatives

- Use Secret instead of ConfigMap. However, inspection data doesn't contain
    confidential information, so there is no benefit we would get by having
    the data encoded.
- Create a new Custom Resource type, and write the data into its spec. Since
    there is no control loop specifically on the inspection data and we only
    write, read or delete it, it seems redundant to introduce a new CRD for

>     this purpose when Kubernetes offers built in objects like ConfigMap/Secret.

Reviewer 2: I disagree here. The value of a CRD is not that you can write a controller to reconcile it (you can write a controller to reconcile anything, including ConfigMaps), it's that the schema is documented, verified, and versioned, and that specific RBAC rules can be applied to it.

I think defining a new CRD that puts the hardware details in the Spec instead of the Status is the correct solution. (This also solves the name collision issue, which also makes it easier to roll out in v1alpha1.)

I documented how I think this should work in [the API decomposition proposal](https://docs.google.com/document/d/1FL8oA0_WNcPdiC-0zrToinl8aol-NPQUjWe-OkcvTqA/edit#heading=h.vb3zlo2h9t7).

Author: @zaneb thanks a lot for the quick review. I agree now that it would make more sense to have a new CRD instead of ConfigMap, which was something I didn't have in my mind while designing. 
If we go with a CRD and keep the hardwareDetails for some time (not to break current users and give enough time for others to prepare), is there some other blocker you can see? I can re-write the proposal, and request another review and after addressing comments, can we right away start implementation or do we have some major plans with regard to API change that this work should hold on?

Reviewer 2: I don't think there's a blocker. We have added a bunch of new CRDs without bumping the version. Removing the existing HardwareDetails would be a blocker, but I think the answer is just to copy them from the CRD if it is available into the Status, and vice-versa, but exclude it from the status annotation. That way it survives the pivot (via the CRD), doesn't break the status annotation size limit, and presents the same behaviour for existing API users.

Author: Awesome. Thanks again, I will ping once I updated the proposal.

Author: Updated accordingly.

# inspectionData custom resource for host inspection data

## Status

implementable

## Summary

> Introduce InspectionData Custom Resource (CR) to store host inspection data.

Author: yup, sounds good to me :+1:

will contain the same hardware information as currently under `status.hardware`
of BareMetalHost.

### Pivoting

There are a couple of options one can choose to pivot objects to a target cluster
that are outside of the Cluster API chain.

1. [`clusterctl.cluster.x-k8s.io/move-hierarchy`](https://github.com/kubernetes-sigs/cluster-api/blob/0e6428151f2a388b98d1e220a698b85940a5f290/cmd/clusterctl/api/v1alpha3/labels.go#L38) label should be set
    on BareMetalHost CRD before issuing *`clusterctl move`* so that the entire hierarchy
    (i.e. including InspectionData, Secrets and ConfigMaps) is pivoted to the target
    cluster.

2. `clusterctl.cluster.x-k8s.io=""` and `cluster.x-k8s.io/provider="metal3"`
    labels on InspectionData CRD to extend the list of provider objects.

3. `clusterctl.cluster.x-k8s.io/move` label on InspectionData CRD. This will
    not move child objects as in option 1.

### Implementation Details/Notes/Constraints

The life of an InspectionData is tightly coupled to a BareMetalHost. Creating or
deleting a BareMetalHost will result in creation or deletion of a InspectionData
respectively. Deprovisioning of the BareMetalHost will result in the inspection

> data removal from the spec of InspectionData.

Author: > I don't think we should bother to create the CR until we do inspection. 

When we create a BareMetalHost, it automatically goes through inspection state until it reaches available right? If yes, why should not we create the CR whenever BareMetalHost is created? 

> We definitely must not remove the inspection data on deprovisioning. 

Sorry my bad, I had a wrong impression that we delete the inspection data when deprovisioning. Though, related to immutability point above, if the CR is immutable, and if someone requests inspection (due to some hardware change) of `Available` BMH, we can't update the CR. So, then we delete and re-create it ?
 
> We should add a finalizer on the CR to prevent it being deleted while a Host is provisioned. More details on how I think the lifecycle should work: https://docs.google.com/document/d/1FL8oA0_WNcPdiC-0zrToinl8aol-NPQUjWe-OkcvTqA/edit#heading=h.vb3zlo2h9t7

+1

> data removal from the spec of InspectionData.

Reviewer 2: > > I don't think we should bother to create the CR until we do inspection.
> 
> When we create a BareMetalHost, it automatically goes through inspection state until it reaches available right? If yes, why should not we create the CR whenever BareMetalHost is created?

Eventually, but there are a couple of states to go through first. If we get stuck in e.g. Registration there is probably no reason to make it look like we've already done inspection by creating the CR. Also, there's an annotation you can add to skip inspection.

> > We definitely must not remove the inspection data on deprovisioning.
> 
> Sorry my bad, I had a wrong impression that we delete the inspection data when deprovisioning. Though, related to immutability point above, if the CR is immutable, and if someone requests inspection (due to some hardware change) of `Available` BMH, we can't update the CR. So, then we delete and re-create it ?

Yes. I think this will make implementation easier too.

> > We should add a finalizer on the CR to prevent it being deleted while a Host is provisioned. More details on how I think the lifecycle should work: https://docs.google.com/document/d/1FL8oA0_WNcPdiC-0zrToinl8aol-NPQUjWe-OkcvTqA/edit#heading=h.vb3zlo2h9t7
> 
> +1

**Note:** [cross-namespace owner references are not allowed](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/#owner-references-in-object-specifications). Thus, InspectionData
and BareMetalHost must be in the same namespace.

- BareMetalHost is created
- Operator creates a InspectionData with the same name and namespace as
    BareMetalHost and sets ownerReferences

> - Provisioning is requested, operator fills out the InspectionData with the

Author: My bad, :+1:

BareMetalHost. Since InspectionData is dependent on BareMetalHost, we need to
ensure that when Cluster is paused, we pause reconciliation of InspectionData too.

The spec of InspectionData:

```go
// Adding HardwareDetails struct here for visibility. During implementation,
// we don't have to duplicate it, but use from baremetalhost_types.go.
type HardwareDetails struct {

    SystemVendor HardwareSystemVendor `json:"systemVendor,omitempty"`
    Firmware     Firmware             `json:"firmware,omitempty"`
    RAMMebibytes int                  `json:"ramMebibytes,omitempty"`
    NIC          []NIC                `json:"nics,omitempty"`
    Storage      []Storage            `json:"storage,omitempty"`
    CPU          CPU                  `json:"cpu,omitempty"`
    Hostname     string               `json:"hostname,omitempty"`

}

type InspectionDataSpec struct {
    // The hardware discovered on the host.
    HardwareDetails *HardwareDetails `json:"hardware,omitempty"`
}
```

The status of InspectionData:

```go
type InspectionDataStatus struct {
    // Last time the status was updated/observed.
    // +optional

>     LastUpdated *metav1.Time `json:"lastUpdated,omitempty"`

Author: I'm +1 on immutability point and based on that it makes sense to drop `LastUpdated`, because there is no possibility anyways updating the CR due to immutability.

>     LastUpdated *metav1.Time `json:"lastUpdated,omitempty"`

Author: ow you meant with RBAC, I was rather thinking to do it with ValidationgWebhook.

>     LastUpdated *metav1.Time `json:"lastUpdated,omitempty"`

Reviewer 2: We can do both :)

}
```

### Risks and Mitigations

> Any suggestions?

Author: maybe I should mention potential security issue here and the need for some RBAC or securing the cluster?

Reviewer 2: I don't know that there's any change in the security implications (if anything it's better).

One risk could be the fact that anybody upgrading an existing metal³ deployment will now need to have installed this CRD and the relevant RBAC permissions.

Author: Good point. I shall include it.

### Work Items

- Add HardwareDetails API
- Modify the operator code
- Clean up statusAnnotation related code from BMO and CAPM3
- Add unit tests
- Update documentation

### Dependencies

None

### Test Plan

Unit tests should be added. Integration test doesn't require any change.

### Upgrade / Downgrade Strategy

From a user perspective, this feature doesn't introduce any breaking changes
to existing deployments, because `status.hardware` field will stay for the
time being and this feature is only an addition to the current APIs. For now,
there will be two copies of inspection data available, in the `spec.hardware`

> of InspectionData and `status.hardware` of BareMetalHost.

Reviewer 2: We should specify somewhere above how we plan to keep them in sync: i.e. copy one to the other if only one exists, new resource takes precedence if they are different.

### Version Skew Strategy

> I can not foresee any version skew issues.

Reviewer 2: I think the strategy is doing a bi-directional copy to keep the BaremetalHost and the new CR in sync. That will prevent any skew issues.

Author: Yup, thank you very much for suggestion

# hostData custom resource for host inspection data

## Status

implementable

## Summary

Introduce HostData Custom Resource (CR) to store host inspection data.

## Motivation

Currently inspection data is written to baremetalhost status subresource. When
pivoting to a target cluster, with *`clusterctl move`*, status subresource is
not copied with the host, because from Cluster-API perspective, it's controllers'
job to rebuild the status in the target cluster. Also, according to the [Kubernetes API convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#spec-and-status), status is expected to be reconstructed by controllers if

> it's lost for some reason. Baremetal Operator (BMO) is capable of reconstructing

Reviewer 3: ```suggestion
it's lost for any reason. Baremetal Operator (BMO) is capable of reconstructing
```

host status by re-inspecting the host, but it will force the already provisioned

> host to go through another round of provisioning, which eventually reboots the

Reviewer 4: I would say another round of inspection

Author: I think it is also important to let people know that the host will be rebooted, which is I think when host is provisioning?

Reviewer 4: yes true

host after IPA steps are completed. To have the inspection data available in the
target cluster and to avoid unnecessary provisioning after pivoting, we currently
copy inspection data to baremetalhost.metal3.io/status annotation so that BMO in

> the target cluster can continue reconciliation from a point it was before pivoting,

Reviewer 4: BMO can reconstruct the BMH in the target cluster from the annotation with the appropriate inspection data and as such the reconciliation of the BMH can continue from the same status as it was in the source cluster.

avoiding redundant provisioning.

However, there is a limitation with the current approach, because when inspection

> data is quite big (e.g. when the host has 1000 multipath disks), it is easy to hit

Reviewer 3: ```suggestion
data is very large (e.g. when the host has 1000 multipath disks), it is easy to hit
```

> the global limit of `metadata.annotations` [262144](https://github.com/kubernetes/kubernetes/blob/master/test/integration/controlplane/synthetic_controlplane_test.go#L313) bytes and fail to pivot successfully.

Reviewer 4: ```suggestion
the global limit of `metadata.annotations` [262144](https://github.com/kubernetes/kubernetes/blob/master/test/integration/controlplane/synthetic_controlplane_test.go#L313) bytes and once the limit is hit the move process will fail to pivot BMH successfully.
```

### Goals

> - Store host inspection data into Spec HostData CR

Reviewer 4: this line is not clear to me.

Author: sorry, but what exactly ?

Author: ah I see, I am missing `of`

- Adjust pivoting workflow to restore inspection data from HostData

> - Make the HostData CR immutable

Reviewer 4: How would it work in case of re-introspection when some new hardware is added for example?

Reviewer 4: I think  I got the answer down below.

### Non-Goals

> Remove `status.hardware` from BareMetalHost CR

Reviewer 4: Until this point I am a bit confused between hostData and inspectionData. If we are only moving `status.hardware` then we are not moving the other parts of baremetalhost.Status data and would that mean that move process has not to take care of the status.hardware from the new CRD and then the rest of the status from annotation as before?

Author: Ah, sorry if you got confused from the PR description, we decided to rename inspectionData to HostData

## Proposal

### User Stories

As an operator, from an ephemeral cluster I would like to pivot/move to a target

> cluster successfully.

Reviewer 4: pivot what? Shall we be more specific since this proposal isn't really targeting the whole of the pivot/move process

## Design Details

Apart from writing hardware details into the status of BareMetalHost, write it to a
HostData CR spec. BareMetalHost and HostData will be parent and child objects
respectively, and this relationship is controlled via ownerReferences chain. As such,
the same name & namespace will be used for HostData as its owning object BareMetalHost.
The spec of HostData will contain the same fields & hardware information as currently
under `status.hardware` of BareMetalHost.

### Pivoting

There are three options one can choose to pivot, objects excluded from Cluster
API chain, to a target cluster:

1. [`clusterctl.cluster.x-k8s.io/move-hierarchy`](https://github.com/kubernetes-sigs/cluster-api/blob/0e6428151f2a388b98d1e220a698b85940a5f290/cmd/clusterctl/api/v1alpha3/labels.go#L38) label should be set
    on BareMetalHost CRD before issuing *`clusterctl move`* so that the entire hierarchy

>     (i.e. including HostData, Secrets and ConfigMaps) is pivoted.

Reviewer 4: I am a bit unsure still if thats the right label although it seems from the link you shared it is. This would need some testing

Author: You are right, and I removed other alternatives not to complicate it here.

2. `clusterctl.cluster.x-k8s.io=""` and `cluster.x-k8s.io/provider="metal3"`

>     labels on HostData CRD to extend the list of provider objects.

Reviewer 4: My initial testing suggests only `clusterctl.cluster.x-k8s.io=""` label is enough but there isn't any harm on having `cluster.x-k8s.io/provider` label. But then it should be set to `infrastructure-metal3` instead of `metal3`.

Author: how did it work for us until now where we were using `metal3` in metal3-dev-env ? Or was it still ok because it was not necessary ?

Author: So, apparently neither infrastructure-metal3 nor metal3 is needed, because of the ownerRefrence chain. So, I will remove the second label.

> 3. `clusterctl.cluster.x-k8s.io/move` label on HostData CRD. This will

Reviewer 4: what are the child objects hostData could have?

Author: hostData itself is a child object to BMH. So, if you set it to BMH, hostData will not be moved.

### Implementation Details/Notes/Constraints

The life of HostData is dependent on BareMetalHost. When BareMetalHost is entering
inspection state, the HostData will be created. Failure in reaching inspection state
or having `inspect.metal3.io: disabled` annotation on BareMetalHost will force the

> operator to not create the HostData. Deleting a BareMetalHost will result in deletion

Author: Good point, I will also mention it in the proposal so that I don't forget it during implementation.

or having `inspect.metal3.io: disabled` annotation on BareMetalHost will force the
operator to not create the HostData. Deleting a BareMetalHost will result in deletion
of a HostData respectively.

> HostData should be immutable and this will be controlled via [Validating Webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) and with RBAC permissions.

Reviewer 5: Can this part be optional?
Running without `Validating Webhook` or with custom `Validating Webhook` should not break HostData CR code.

Author: Is it so that you want to see the CR to be mutable and if yes may I ask why?

Reviewer 5: No, was pointing out that someone could have custom `Validating Webhook` already in place, this could break immutability assumption.
One use case for custom web-hook is mutating webhook, to filter out some of unnecessary data from HostData CR.

So we should probably rephrase `HostData should be immutable` to something like `HostData should be protected from unintentional change by Validating Webhook`

In case, when [re-inspection](https://github.com/metal3-io/baremetal-operator/blob/master/docs/inspectAnnotation.md) of a `available` BareMetalHost is requested,
new HostData will be re-created, because updating the old HostData will be
blocked by the Validating Webhook.

Deprovisioning of the BareMetalHost must not delete the HostData. It can
be deleted either when BareMetalHost is being deleted or when re-inspection
is requested. To ensure HostData availability while BareMetalHost is provisioned,
a finalizer should be set on the HostData to prevent its accidental deletion.

Only operator will be granted with create permissions for HostData and everyone
else will be given no write permissions. If the BareMetalHost was inspected but
HostData object doesn't exist, operator will first try to copy it from the

> BareMetalHost status (if available) and vice versa. If not available in neither

Reviewer 4: This point is not clear. Once hostData exists  will it be available in BMH status anymore?

Author: if the BMH status doesn't have inspection data, but hostData has, then we will copy it from hostData to the status

> of them, new inspection will be enforced.

Reviewer 6: A finalizer can always be manually patched to force a deletion at any time. This could generate a situation where an already provisioned (or any other state after inspection) host could suddenly loose the inspection data. Forcing re-inspection could become tricky (and this could be more problematic when the field will be removed from the host status)

Author: That's right. However, I think in general we give cluster access to someone we trust. For example, accidental deletion of BMH can also happen if the cluster access is not secured. So, perhaps we will need to clearly mention that in the user document as potential cause of not securing the cluster?

Also, that's how Cluster API workflow is also designed. For example, if someone intentionally/accidentally deletes Cluster object, or Machine, then it will result in other objects deletion/deprovisioning on the chain.

Reviewer 6: More than a security problem I think that probably the `new inspection will be enforced` phrase could be detailed a little bit more. In general, if the HostData deletion will happen after the `Inspecting` and within the `Available` state, the host could be forced back to the `Inspecting` without any particular issue. Instead, if it happens during a `Provisioned` state, as you also already mentioned, this could  trigger a deprovisioning first

of them, new inspection will be enforced.

#### Keeping inspection data in sync

> As mentioned above, we will have two copies of inspection data available. One in the Spec

Reviewer 7: Maybe I missed it above, but I don't quite understand why keep two copies of a quite large structure.

of HostData and the other in the Status of BareMetalHost. One of them will get its inspection
data copied from the other. Thus, inspection data can be copied from HostData (if available)
to BareMetalHost, or the vice versa. If there is delta in the inspection data while

> copying, HostData should take precedence and get copied to the Status of BareMetalHost.

Reviewer 4: I would really make a sync between them whenever the pause annotation is applied so that we dont have any mismatched status before and after pivot.

Author: @andfasano yes that's right. This double source of inspection data management will get much easier in the future when we drop the `status.hardware` from BMH.

Reviewer 6: I'm wondering at this point if it could be better to remove the `vice versa` part, in order to also to prepare the ground for the future drop of `status.hardware` - I understand that this could mean a more complex implementation as of now.

Author: Sure, I'm more than happy to consider hostData as the main source of truth for the inspection data. As you mentioned, while `status.hardware` still exists, it can serve as a backup for the hostData only when hostData needs to be recreated. 

I remember @zaneb suggesting making it clear how we keep hostData and `status.hardware` of BMH in sync. So, I hope @zaneb would also agree on this?
To clarify the idea is:
- HostData is the main source of inspection data;
- Inspection data in the hostData can be populated/copied from BMH status only if hostData is missing
- Until we remove the `status.hardware` from BMH, inspection data can be populated/copied from the hostData to the status of BMH in case it is missing. This part is temporary, and will be removed during the API bump.

Reviewer 6: As also mentioned in a later [comment](https://github.com/metal3-io/metal3-docs/pull/208#discussion_r773191847), a new problem to handle is how to manage the case where HD suddenly disappears (maybe due a deletion). In any case I would not use the BMH status to regenerate it, either because HD will be the single source of truth, either because it be removed in future

and BareMetalHost must be in the same namespace.

- BareMetalHost is created
- BareMetalHost is in inspecting state
- Operator creates a HostData with the same name and namespace as BareMetalHost, fills out
    the HostData with the host inspection data and sets ownerReferences
- Provisioning is requested
- Pivoted to a target cluster, operator restores host inspection data from HostData and
    the rest from status annotation (e.g. `status.Provisioning`, `status.poweredOn` and etc)
- Deprovisioning is requested
- BareMetalHost deletion requested, operator deletes the HostData and BareMetalHost

> The spec of HostData:

Reviewer 6: It could be useful to have the complete to CR, ie to see also the ownerRef

```go
// Adding HardwareDetails struct here for visibility. During implementation,
// we don't have to duplicate it, but use from baremetalhost_types.go.
type HardwareDetails struct {

    SystemVendor HardwareSystemVendor `json:"systemVendor,omitempty"`
    Firmware     Firmware             `json:"firmware,omitempty"`
    RAMMebibytes int                  `json:"ramMebibytes,omitempty"`
    NIC          []NIC                `json:"nics,omitempty"`
    Storage      []Storage            `json:"storage,omitempty"`
    CPU          CPU                  `json:"cpu,omitempty"`
    Hostname     string               `json:"hostname,omitempty"`

}

type HostDataSpec struct {
    // The hardware discovered on the host.
    HardwareDetails *HardwareDetails `json:"hardware,omitempty"`
}
```

### Risks and Mitigations

Existing deployments will require installation of the HostData CRD and necessary
RBAC permissions when bumping the operator version. This needs to be clearly
documented.

### Work Items

- Add HostData API
- Modify the operator code
- Clean up statusAnnotation related code from BMO and CAPM3
- Add unit tests
- Update documentation

### Dependencies

None

### Test Plan

> Unit tests should be added. Integration test doesn't require any change.

Reviewer 4: The pivot test which is a part of the integration test should have a change right?

Author: nope, it doesn't change the way clusterctl pivot works nor the way we trigger the pivoting

> Unit tests should be added. Integration test doesn't require any change.

Author: AFAIK we don't yet have e2e in BMO. If there is one in the future I can add it, but perhaps for now this can be continued to be tested in daily master jobs (Jenkins) that use metal3-dev-env code base?

> Unit tests should be added. Integration test doesn't require any change.

Reviewer 6: That's right, in any case I think there could be some value to define, even in a pure natural language, which e2e test could make sense (or it's reasonable) to have for validating/accepting the new feature - so that maybe it could be easier in future to implement them having previously defined in the design doc. I wasn't pointing to an exhaustive or super-detailed list, just a few of them

### Upgrade / Downgrade Strategy

From a user perspective, this feature doesn't introduce any breaking changes

> to existing deployments, because `status.hardware` field will stay until is

Reviewer 7: Okay, this is why we may want to keep the old location... Do we need a plan to deprecate it? We should do it before the BMH API reaches beta.

Reviewer 2: Yes, our next (first) API version bump should delete this field from the BMH.

decided otherwise and this feature is only an addition to the current APIs.
For now, there will be two copies of inspection data available, in the
`spec.hardware` of HostData and `status.hardware` of BareMetalHost.

### Version Skew Strategy

To mitigate version skew issues, we plan to keep BaremetalHost and HostData
synchronized by bi-directional copying of inspection data.

## Drawbacks

None

## Alternatives

- Use k8s built in objects like ConfigMap/Secret. From security point of view, although,
    it is possible to restrict permissions to edit ConfigMaps (only to those
    storing inspection data) based on the name (i.e. `resourceNames`), but it will add
    complexity to the operator.

> - Increase metadata global limit. However, this does not always work as it is hard

Reviewer 5: Adding filters to filter-out unneeded inspection data?

Author: Thanks for suggestion, but I'm afraid that even if we add a filter to ignore unnecessary fields we still will hit the issue, because the size of the unnecessary fields is relatively small and instead it is `status.hardware` that we care about and which might grow a lot depending on the hardware.

Reviewer 5: valid point
